### PR TITLE
Use DistributedSampler when running with custom accelerator

### LINKED ISSUE #7813: DistributedSampler not used when using DDPPlugin
## 🐛 Bug

When running with DDPPlugin, the DataLoaders are not configured with a DistributedSampler.

### To Reproduce

Run [this code](https://gist.github.com/ajtritt/934f82438c582e9a7e8e7cd9c335ae05) in an LSFEnvironment. Regardless of the number of gpus specified (i.e. the first argument), the number of batches is always 3750. After doing some digging, it looks like [auto_add_sampler](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/data_loading.py#L113) is not setting the sampler correctly.

### Expected behavior

When running with DDP training, the sampler for the DataLoaders should be set to a DistributedSampler. 

### Environment

 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): RedHat
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): python setup.py develop
 - Python version: 3.8
 - CUDA/cuDNN version:
 - GPU models and configuration: 1 or 6 GPUs
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by ajtritt**
## What does this PR do?

Fixes #7813.

This only fixes the problem for DDDPlugin. I suspect the same problem exists for other distributed data parallel plugins, but I'm not familiar enough with those training types to verify that. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -403,7 +403,11 @@ def select_precision_plugin(self) -> PrecisionPlugin:
         raise NotImplementedError("We only support precisions 64, 32 and 16!")
 
     def select_training_type_plugin(self) -> TrainingTypePlugin:
-        if self.use_ddp2:
+        if isinstance(
+            self.distributed_backend, Accelerator
+        ) and self.distributed_backend.training_type_plugin is not None:
+            plugin = self.distributed_backend.training_type_plugin
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:409]: Nice !

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:409]: Mind adding a test ?

Comment (ajtritt) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:409]: done.

File: tests/trainer/connectors/test_accelerator_connector.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import torch
+
+from pytorch_lightning import Trainer
+from pytorch_lightning.accelerators import Accelerator
+from pytorch_lightning.plugins import DDPPlugin, PrecisionPlugin, SingleDevicePlugin
+
+
+def test_accelerator_training_type_plugin():
```
Comment (awaelchli) [inline tests/trainer/connectors/test_accelerator_connector.py:8]: can you move this test to `tests/accelerators/test_accelerator_connector` and integrate it with `test_custom_accelerator`?

Comment (ajtritt) [reply tests/trainer/connectors/test_accelerator_connector.py]: done

**Review by tchaton — APPROVED** [review]
LGTM ! Nice catch !

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:409]: do we need to add the same fix to the `select_precision_plugin()` method?

**Review by awaelchli — APPROVED** [review]
we probably have to add the same fix to `select_precision_plugin()`?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -403,7 +403,11 @@ def select_precision_plugin(self) -> PrecisionPlugin:
         raise NotImplementedError("We only support precisions 64, 32 and 16!")
 
     def select_training_type_plugin(self) -> TrainingTypePlugin:
-        if self.use_ddp2:
+        if isinstance(
+            self.distributed_backend, Accelerator
+        ) and self.distributed_backend.training_type_plugin is not None:
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:408]:
```suggestion
        is_plugin = self.distributed_backend.training_type_plugin is not None
        if isinstance(self.distributed_backend, Accelerator) and is_plugin:
```

**Review by Borda — APPROVED** [review]

**Comment by tchaton** [comment]
Dear @ajtritt,

The tests are failing. Do you need help resolving them ?

Best,
T.C

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -403,7 +403,10 @@ def select_precision_plugin(self) -> PrecisionPlugin:
         raise NotImplementedError("We only support precisions 64, 32 and 16!")
 
     def select_training_type_plugin(self) -> TrainingTypePlugin:
-        if self.use_ddp2:
+        is_plugin = self.distributed_backend.training_type_plugin is not None
+        if isinstance(self.distributed_backend, Accelerator) and is_plugin:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:407]: this has to be merged into the if statement. only the Accelerator object has the training_type_plugin attribute.

```suggestion
        if isinstance(self.distributed_backend, Accelerator) and self.distributed_backend.training_type_plugin is not None:
```

**Comment by awaelchli** [comment]
@ajtritt Thanks for looking into this. Are you interested in sending an equivalent fix for the precision plugin logic for a custom accelerator?
